### DISCUSSION TITLE
[v3] Need an additional event before executing the updateCurrentPageHtmlInHistoryStateForLaterBackButtonClicks method

### DISCUSSION THREAD

**Discussion by N1ebieski** [livewire/livewire#5894] [1. Help] [answered]
Livewire v3 has a great feature that is called "wire:navigate". Unfortunately, this feature has one big problem, doesn't use the browser cache for the back and forward buttons. Instead it uses own system. First it creates a timestamp key and saves it in the history.state._html, then saves entire document.documentElement.outerHTML to sessionStorage for this key. If user clicks back button, Livewire injects entire html document from the cache.

In most cases this is enough, ~~for example all Alpine components will work properly~~. But not vanilla javascript components that change an element in the DOMDocument and store their state in it.

Let me explain with an example [Tomselect](https://github.com/orchidjs/tom-select), a popular implementation select2:

```html
            <select
                x-data="{ value: $wire.$entangle('select') }"
                x-init="
                    const tomselect = new TomSelect($refs.select, {
                        labelField: 'null'
                    });
                    tomselect.on('change', (value) => {
                        this.value = value;
                    });
                "
                class="form-select"
                wire:model="select"
                wire:ignore
                x-ref="select"
                x-cloak                
            >
                <option value="one">One</option>
                <option value="two">Two</option>
            </select>
```

initialization of TomSelect by Alpine will replace the normal html select element in DOMDocument from:

```html
            <select
                class="form-select"
            >
                <option value="one">One</option>
                <option value="two">Two</option>
            </select>
```

to:

```html
<select x-data="{ value: $wire.$entangle('select') }" x-init="
                    const tomselect = new TomSelect($refs.select, {
                        labelField: 'null'
                    });
                    tomselect.on('change', (value) => {
                        this.value = value;
                    });
                " class="form-select tomselected ts-hidden-accessible" wire:model="select" wire:ignore="" x-ref="select" id="tomselect-1" tabindex="-1">
                
                <option value="two">Two</option>
                <option value="one">One</option>
</select>
<div class="ts-wrapper form-select single input-hidden full has-items">
   <div class="ts-control"><div data-value="one" class="item" data-ts-item="">One</div>
      <input type="select-one" autocomplete="off" size="1" tabindex="0" role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-controls="tomselect-1-ts-dropdown" id="tomselect-1-ts-control">
   </div>
   <div class="ts-dropdown single" style="display: none;">
      <div role="listbox" tabindex="-1" class="ts-dropdown-content" id="tomselect-1-ts-dropdown"></div>
   </div>
</div>
```

In addition, it will save the state of the entire object to $refs.select.tomselect:

![tomselect](https://github.com/livewire/livewire/assets/14810714/d5389a25-de68-4455-84cc-863215bec7cd)

Everything will work fine until you go to another page using wire:navigate and then return using the browser's back button.

Then Livewire will recreate the above component code, but without the previous state $refs.select.tomselect! Alpine will try to initiate the component again, but then it will duplicate the above code.

So what solution do I see?

We need an additional event before executing the [updateCurrentPageHtmlInHistoryStateForLaterBackButtonClicks](https://github.com/alpinejs/alpine/blob/a9206161f0bf878d41a001197553613cf31cbde9/packages/navigate/src/history.js#L2-L8) method which stores document.documentElement.outerHTML to sessionStorage. It would then be possible to restore the code of such components to their original state (In the case of TomSelect, it is enough to call the method $refs.select.tomselect.destroy() ).

[Turbo Wire](https://turbo.hotwired.dev/reference/events), predecessor "wire:navigate" has such an event. it's called "turbo:before-cache" and fires before Turbo saves the current page to cache.

**Comment by N1ebieski** [comment]
Another simplest example (normal Alpine component):

```html
            <div 
                x-data="{ values: [1, 2] }"
            >
                <ul>
                    <template x-for="(value, index) in values" :key="index">
                        <li x-text="value"></li>
                    </template>
                </ul>
                <button x-on:click.prevent="values.push(values[values.length - 1] + 1)">
                    Add
                </button>
            </div>
```

[screen-recording.webm](https://github.com/livewire/livewire/assets/14810714/7383d4be-7677-465e-af9b-8c86f2e2996f)

**Comment by guillaumelebon** [comment]
After some testing with `wire:navigate`, I'm encountering similar issues with the browser's back and forward buttons.

It seems that the problems stem from the fact that Livewire attempts to cache the state of each page, allowing for quicker navigation to previous or next pages.

I'm not sure if this behavior is intended for the "SPA" feeling, but in my view it makes `wire:navigate` unusable for most use cases.

A very standard situation:

- `wire:navigate` to a list of items
- `wire:navigate` to a specific item named "foo"
- change the item name to "bar"
- press browser's back button

With the current caching strategy, the selected item name is still named "foo".

Generally speaking, a lot can happen between the time a user leaves a page and the time they return to it.

In most use cases I see, I expect to have a "fresh" state when I change a page.

What After some testing with `wire:navigate`, I'm encountering similar issues with the browser's back and forward buttons.

It seems that the problems stem from the fact that Livewire attempts to cache the state of each page, allowing for quicker navigation to previous or next pages.

I'm not sure if this behavior is intended for the "SPA" feeling, but in my view it makes `wire:navigate` unusable for most use cases.

A very standard situation:

- `wire:navigate` to a list of items
- `wire:navigate` to a specific item named "foo"
- change the item name to "bar"
- press browser's back button

With the current caching strategy, the selected item name is still named "foo".

Generally speaking, a lot can happen between the time a user leaves a page and the time they return to it.

In most use cases I see, I expect to have a "fresh" state when I change a page.

What are your thoughts on this?

**Comment by calebporzio** [comment, marked as answer]
Heads up, the next beta release will have a `wire:navigating` event that can be used to mutate the HTML of the page before navigating away so that you have full control over what goes in the back button cache

**Reply by N1ebieski** [reply]
Thank Caleb. The event is called "wire:navigating" or "livewire:navigating"? :)

Correct in the [documentation](https://livewire.laravel.com/docs/navigate#updating-the-page-before-navigating-away)
